core/src/main/scala/kafka/controller/KafkaController.scala

```diff
@@ -1425,25 +1589,134 @@ class KafkaController(val config: KafkaConfig,
+
+      case None => // ZK triggered
+        // We need to register the watcher if the path doesn't exist in order to detect future reassignments and we get
+        // the `path exists` check for free
+        if (zkClient.registerZNodeChangeHandlerAndCheckExistence(partitionReassignmentHandler)) {
+          val partitionReassignment = zkClient.getPartitionReassignment
+          val (savedReassignments, _) = partitionReassignment.partition { case (tp, targetReplicas) =>
+            savePartitionReassignment(reassignmentResults, tp, Some(targetReplicas), zkTriggered = true)
+          }
+          savedReassignments.keySet
+        } else {
+          Set.empty[TopicPartition]
+        }
+    }
 
-    // We need to register the watcher if the path doesn't exist in order to detect future reassignments and we get
-    // the `path exists` check for free
-    if (zkClient.registerZNodeChangeHandlerAndCheckExistence(partitionReassignmentHandler)) {
-      val partitionReassignment = zkClient.getPartitionReassignment
+    reassignmentResults ++= maybeTriggerPartitionReassignment(partitionsToBeReassigned)
+    callback match {
+      case Some(cb) => cb(Left(reassignmentResults))
+      case None =>
+    }
+  }
 
-      // Populate `partitionsBeingReassigned` with all partitions being reassigned before invoking
-      // `maybeTriggerPartitionReassignment` (see method documentation for the reason)
-      partitionReassignment.foreach { case (tp, newReplicas) =>
-        val reassignIsrChangeHandler = new PartitionReassignmentIsrChangeHandler(eventManager, tp)
-        controllerContext.partitionsBeingReassigned.put(tp, ReassignedPartitionsContext(newReplicas, reassignIsrChangeHandler))
-      }
+  private def replicasAreValid(replicasOpt: Option[Seq[Int]]): Boolean = {
+    replicasOpt match {
+      case Some(replicas) =>
+        val replicaSet = replicas.toSet
+
+        if (replicas.size != replicaSet.size)
+          false
+        else if (replicas.exists(_ < 0))
+          false
+        else
+          replicaSet.subsetOf(controllerContext.liveBrokerIds)
+      case None => true
+    }
+  }
+
+  private def savePartitionReassignment(reassignmentResults: mutable.Map[TopicPartition, ApiError], partition: TopicPartition,
+                                        targetReplicasOpt: Option[Seq[Int]], zkTriggered: Boolean): Boolean = {
+    val reassignIsrChangeHandler = new PartitionReassignmentIsrChangeHandler(eventManager, partition)
+    val replicaAssignment = controllerContext.partitionFullReplicaAssignment(partition)
+    val reassignmentIsInProgress = controllerContext.partitionsBeingReassigned.contains(partition)
+
+    val newContextOpt = targetReplicasOpt match {
+      case Some(targetReplicas) =>
+        if (reassignmentIsInProgress) {
+          info(s"Overriding old reassignment for partition $partition " +
+            s"(with target replicas ${replicaAssignment.targetReplicas.mkString(",")}) " +
```

Comment:
nit: can `replicaAssignment.targetReplicas` be different than `targetReplicas`? If not this log statement might be a bit redundant.

Reply:
Yes, that's the reason of the log. It is called when we are overriding an existing reassignment. i.e an on-going reassignment is happening (`replicaAssignment`) and we call the function to save a new reassignment (`targetReplicas`)

They can be the same, sure, if the user sends the same request twice in a row - but it is worth logging regardless. At the very least, it would help diagnose bugs easier. Besides that, I think it's a very useful log